src/persist-client/src/read.rs

```diff
@@ -1103,23 +1116,31 @@ where
             self.cfg.dynamic.compaction_memory_bound_bytes(),
         );
 
+        let metadata = SerdeLeasedBatchPartMetadata::Snapshot {
```

Comment:
I think all of this goes away if we use `self.snapshot` rather than `self.machine.snapshot` at the start of this method -- that'll grab a snapshot and lease all the parts it returns. I think the original bug comes from using this latter method, which doesn't place a seqno hold, so individual parts can be deleted from under the consolidator (even if the `since` doesn't advance the parts could change)

Reply:
Unfortunately we can't use `self.snapshot` here, since it loses the run structure of the shard.

An alternative would be to rewrite `self.snapshot` to handle all the runs separately, but that's a more invasive change. So far I have been trying to not rewrite existing code to keep the footprint small, but I'm open to it if you prefer it!

Reply:
Ah right! Okay then this all checks out to me

> I'm a little unclear why this could cause failures in snapshot_and_fetch_streaming, since we're holding a unique ref to our handle for the full call, which I would have thought would prevent us from advancing the since too far ahead. But in any case it seems like the right thing is to match the behaviour of the non-streaming path, and the non-streaming path leases.

Anyway this is from lack of seqno hold, not since, so the change here should fix the bug you ran into

Reply:
> Anyway this is from lack of seqno hold, not since, so the change here should fix the bug you ran into

Yeah... so far the only place I've spotted where we advance the seqno for a reader is during downgrade-since, but I must just be missing something, since the symptoms are pretty clear!

Reply:
That's correct. The seqno that's advanced there is derived from leased parts